### PR TITLE
Make optimiser settings available to assembly stack.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libyul/AssemblyStack.cpp [hunk 1]
```diff
@@ -135,13 +141,15 @@ void AssemblyStack::optimize(Object& _object)
 	for (auto& subNode: _object.subObjects)
 		if (auto subObject = dynamic_cast<Object*>(subNode.get()))
 			optimize(*subObject);
-	// TODO: Store this as setting - it should be the same as the flag passed to
```
Comment (chriseth) [inline libyul/AssemblyStack.cpp:138]: This basically solves this "TODO"

File: libyul/AssemblyStack.cpp [hunk 2]
```diff
@@ -135,13 +141,15 @@ void AssemblyStack::optimize(Object& _object)
...
+		languageToDialect(m_language, m_evmVersion),
+		*_object.code,
+		*_object.analysisInfo,
+		m_optimiserSettings.optimizeStackAllocation
+	);
 }
 
-MachineAssemblyObject AssemblyStack::assemble(Machine _machine, bool _optimize) const
+MachineAssemblyObject AssemblyStack::assemble(Machine _machine) const
 {
```
Comment (bshastry) [inline libyul/AssemblyStack.cpp:153]: `test/tools/ossfuzz/strictasm_assembly_ossfuzz.cpp` needs to be adapted to invoke this API correctly.

File: test/libyul/ObjectCompilerTest.cpp [hunk 3]
```diff
@@ -64,17 +64,20 @@ ObjectCompilerTest::ObjectCompilerTest(string const& _filename)
 
 bool ObjectCompilerTest::run(ostream& _stream, string const& _linePrefix, bool const _formatted)
 {
-	AssemblyStack stack(EVMVersion(), AssemblyStack::Language::StrictAssembly);
+	AssemblyStack stack(
+		EVMVersion(),
+		AssemblyStack::Language::StrictAssembly,
+		m_optimize ? OptimiserSettings::full() : OptimiserSettings::minimal()
+	);
 	if (!stack.parseAndAnalyze("source", m_source))
```
Comment (bshastry) [inline test/libyul/ObjectCompilerTest.cpp:72]: Multiple test harnesses in `test/tools/ossfuzz/` need to be adapted to reflect this change.

**Review by bshastry — CHANGES_REQUESTED** [review]
ossfuzz test harnesses need to be adapted.

**Comment by chriseth** [comment]
@bshastry just updated - can you check if I got everything?

File: solc/CommandLineInterface.cpp [hunk 4]
```diff
@@ -1239,12 +1239,16 @@ bool CommandLineInterface::assemble(
 	map<string, yul::AssemblyStack> assemblyStacks;
 	for (auto const& src: m_sourceCodes)
 	{
-		auto& stack = assemblyStacks[src.first] = yul::AssemblyStack(m_evmVersion, _language);
+		auto& stack = assemblyStacks[src.first] = yul::AssemblyStack(
+			m_evmVersion,
+			_language,
+			_optimize ? OptimiserSettings::full() : OptimiserSettings::minimal()
```
Comment (leonardoalt) [inline solc/CommandLineInterface.cpp:1245]: Why is it `full` here and `standard` in `Common.cpp`?

Comment (chriseth) [reply solc/CommandLineInterface.cpp:1245]: Because this is assembly mode and `--optimize` here means to optimize including yul.

**Comment by bshastry** [comment]
> @bshastry just updated - can you check if I got everything?

Missing change to `test/tools/ossfuzz/yulProto_diff_ossfuzz.cpp`, others are fine.

**Comment by chriseth** [comment]
Rebased.

**Review by bshastry — DISMISSED** [review]

**Comment by chriseth** [comment]
Commandline tests are failing.

**Pull request merged**